webroot/src/pages/FinalizePrivilegePurchase/FinalizePrivilegePurchase.ts

```diff
@@ -0,0 +1,543 @@
+                label: this.stateText,
+                shouldHideLabel: false,
+                shouldHideMargin: true,
+                placeholder: computed(() => this.$t('common.select')),
+                validation: Joi.string().required().messages(this.joiMessages.string),
+                value: '',
+                valueOptions: this.stateOptions,
+            }),
+            zip: new FormInput({
+                id: 'zip-code',
+                name: 'zip-code',
+                label: this.zipLabel,
+                shouldHideLabel: false,
+                shouldHideMargin: true,
+                placeholder: '00000',
+                autocomplete: 'postal-code',
+                shouldHideErrorMessage: true,
+                validation: Joi.string().required().regex(new RegExp('(^[0-9]{5}$)|(^[0-9]{5}-[0-9]{4}$)')),
+            }),
+            submit: new FormInput({
+                isSubmitInput: true,
+                id: 'submit',
+            }),
+        });
+        this.watchFormInputs();
+    }
+
+    async handleSubmit() {
+        this.validateAll({ asTouched: true });
+
+        if (this.isFormValid) {
+            this.startFormLoading();
+            this.isFormError = false;
+            this.formErrorMessage = '';
+
+            const { formValues, statesSelected } = this;
+            const serverData = LicenseeUserPurchaseSerializer.toServer({ formValues, statesSelected });
+            const purchaseServerEvent = await this.$store.dispatch('user/postPrivilegePurchases', serverData);
+
+            this.endFormLoading();
+
+            if (purchaseServerEvent?.message === 'Successfully processed charge') {
+                this.$router.push({
+                    name: 'PurchaseSuccessful',
+                    params: { compact: this.currentCompactType }
+                });
+            } else if (purchaseServerEvent?.message) {
+                this.isFormError = true;
+                this.formErrorMessage = purchaseServerEvent?.message;
+            }
+        } else {
+            this.isFormError = true;
+            this.formErrorMessage = this.formValidationErrorMessage;
+        }
+    }
+
+    handleCancelClicked() {
+        if (this.currentCompactType) {
+            this.$router.push({
+                name: 'LicenseeDashboard',
+                params: { compact: this.currentCompactType }
+            });
+        }
+    }
+
+    handleBackClicked() {
+        if (this.currentCompactType) {
+            this.$router.push({
+                name: 'SelectPrivileges',
+                params: { compact: this.currentCompactType }
+            });
+        }
+    }
+
+    handleExpYearRefEmitted(inputData) {
+        this.expYearRef = inputData.ref;
+    }
+
+    handleCVVRefEmitted(inputData) {
+        this.cvvRef = inputData.ref;
+    }
+
+    togglePaymentCollapsed() {
+        this.shouldShowPaymentSection = !this.shouldShowPaymentSection;
+    }
+
+    formatCreditCard(): void {
```

Comment:
Nice, in the future I would likely prefer to use a library.

I just pushed up some tests for the formatting of the credit card. Because the tests only recreate a virtual dom + we have to interact with it differently than a user would the real DOM, in order to test the functionality, the tests are valuable but less than a perfect 1 to 1. Once we have the mockAPI routeguarding implemented we could consider adding in the usage of cypress of other tools that can better mimic exactly what we want to test here.